src/QsCompiler/Tests.Compiler/TestCases/LinkingTests/Generics.qs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/// This namespace contains test cases for generic operations
+namespace Microsoft.Quantum.Testing.Generics {
+
+	operation Test1Main() : Unit {
+		body (...) {
+			let temp = BasicGeneric<_, Double>(1, _);
+			let temp2 = BasicGeneric<Int, Double>(1, _);
+
+			(BasicGeneric("Yes", _))(1.0);
+
+			(BasicGeneric(BasicGeneric("Yes","No"), _))(BasicGeneric(1.0, 2));
+
+			(BasicGeneric<_,_>((BasicGeneric<_,String>("Yes",_))("No"), _))((BasicGeneric<Double, _>(_, 2))(1.0));
+
+			BasicGeneric((),());
+			BasicGeneric("","");
+			BasicGeneric(1.0,2);
+
+			let temp3 = NoArgsGeneric<Double>();
+			let bar = (ReturnGeneric<String, _, _>("Yes", (ReturnGeneric<Double, _, _>(12.0, "No", _))(4), _))("Maybe");
+		}
+	}
+
+	// Tests that unused generics are removed
+	operation NotUsed<'A, 'B>(a: 'A, b: 'B) : Int {
+		body (...) {
+			using (q = Qubit()) {
+				let temp = a;
+				let temp2 = b;
+			}
+			return 12;
+		}
+	}
+
+	operation BasicGeneric<'A, 'B>(a: 'A, b: 'B) : Unit {
+		body (...) {
+			let temp = a;
+			using (q = Qubit()) {
+				let temp2 = b;
+			}
+		}
+	}
+
+	operation NoArgsGeneric<'A>() : 'A {
+		body (...) {
+			let temp = new 'A[1];
+			return temp[0];
+		}
+	}
+
+	operation ReturnGeneric<'A,'B,'C>(a: 'A, b: 'B, c: 'C) : 'C {
+		body (...) {
+			let temp = c;
+			using (q = Qubit()) {
+				let temp2 = b;
+				let temp3 = a;
+			}
+			return temp;
+		}
+	}
+
+	operation Test2Main () : Unit is Adj+Ctl {
+        body (...) {
+			using (q = Qubit()) {
+				GenericCallsGeneric(q, 12);
+				let temp = ArrayGeneric(q, ArrayGeneric(q, "twelve"));
+				let temp2 = ArrayGeneric(q, 12);
+			}
+        }
+        
+        adjoint (...) {
```

Comment:
I've got an interesting case to check & add a test for: Does the transformation work as expected if a particular type specialization is only invoked as part of the adjoint specialization (or controlled, or controlled adjoint) of a used operation? In particular with additional control qubits, this seems like a valid use case. Since you may not know whether or not a particular functor specialization is executed at compile time, it seems ok to assume that all of the defined functor specializations for a particular set of type arguments will be executed. 
I am happy for this to be a separate PR. ;)